### PR TITLE
Improve status check of model

### PR DESCRIPTION
**Pull request by sunlin-link**

### CODE REVIEW

**Comment by sunlin-link** [comment]
Fixes:
https://github.com/intel/webml-polyfill/issues/77
https://github.com/intel/webml-polyfill/issues/89

which is regression by
https://github.com/otcshare/chromium-src/pull/20

**Comment by sunlin-link** [comment]
When mode is finished, `model_info_` is empty. So `is_finished_` and `model_info_` should not be checked together. The code as below will cause crash.

```
if (InvalidState(is_finished_,
                          index >= model_info_->operands.size() ? "Index is invalid." : "",
                          exception_state))
```

**Comment by sunlin-link** [comment]
@fujunwei Please help to review

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 1]
```diff
@@ -262,4 +243,33 @@ void Model::OnConnectionError() {
...
+  exception_state.ThrowDOMException(
+      DOMExceptionCode::kInvalidStateError, "Model is finished");
+  return true;
+}
+
+bool Model::CheckOperation(const Vector<uint32_t>& inputs,
+                           const Vector<uint32_t>& outputs,
+                           ExceptionState& exception_state) {
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    if (inputs[i] > model_info_->operands.size()) {
```
Comment (sunlin-link) [inline third_party/blink/renderer/modules/ml/model.cpp]: @huningxin Here should be `>` or `>=`?

Comment (huningxin) [reply third_party/blink/renderer/modules/ml/model.cpp]: should be `>=`, thanks for the catching!

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 2]
```diff
@@ -66,11 +38,14 @@ Model::~Model() = default;
 
 void Model::addOperand(const OperandOptions& options,
                        ExceptionState& exception_state) {
-  if (InvalidState(is_finished_,
-                   !options.hasType() ? "Operand type is missing." : "",
-                   exception_state))
+  if (CheckFinished(exception_state))
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: if (InvalidState(!model_info_ ? "Invalid state" :
                   !options.hasType() ? "Operand type is missing." : "",	
                   exception_state))	
    return;

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 3]
```diff
@@ -81,11 +56,13 @@ void Model::addOperand(const OperandOptions& options,
 void Model::setOperandValue(uint32_t index,
                             MaybeShared<DOMArrayBufferView> data,
                             ExceptionState& exception_state) {
-  if (InvalidState(
-          is_finished_,
-          index >= model_info_->operands.size() ? "Index is invalid." : "",
-          exception_state))
+  if (CheckFinished(exception_state))
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: `if (InvalidState(!model_info_ ? "Invalid state" : 
                        index >= model_info_->operands.size() ? "Index is invalid." : "")
    return;  
`

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 4]
```diff
@@ -130,8 +107,10 @@ void Model::addOperation(int32_t type,
                          Vector<uint32_t>& inputs,
                          Vector<uint32_t>& outputs,
                          ExceptionState& exception_state) {
-  if (InValidParameters(is_finished_, inputs, outputs,
-                        model_info_->operands.size(), exception_state))
+  if (CheckFinished(exception_state))
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: `if (InValidParameters(model_info_, inputs, outputs, exception_state))
   return;
`

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 5]
```diff
@@ -18,39 +18,11 @@
...
 namespace {
-bool InvalidState(bool is_finished,
-                  const String& message,
-                  ExceptionState& exception_state) {
-  String error_message = is_finished ? "Model is finished." : "";
-  if (message.IsEmpty() && error_message.IsEmpty())
-    return false;
 
+void InvalidState(ExceptionState& exception_state,
+                  const String& message) {
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: if (message.IsEmpty()) return;

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 6]
```diff
@@ -262,4 +243,33 @@ void Model::OnConnectionError() {
...
+bool Model::CheckFinished(ExceptionState& exception_state) {
+  if (!is_finished_)
+    return false;
+
+  exception_state.ThrowDOMException(
+      DOMExceptionCode::kInvalidStateError, "Model is finished");
+  return true;
+}
+
+bool Model::CheckOperation(const Vector<uint32_t>& inputs,
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: It's better host it in unnamespace with named InvalidParameters.

File: third_party/blink/renderer/modules/ml/model.cpp [hunk 7]
```diff
@@ -262,4 +243,33 @@ void Model::OnConnectionError() {
...
+                           const Vector<uint32_t>& outputs,
+                           ExceptionState& exception_state) {
```
Comment (fujunwei) [inline third_party/blink/renderer/modules/ml/model.cpp]: `if (!model_info_)
  return InvalidState(exception_state, "Invalid state.");
`

**Comment by sunlin-link** [comment]
duplicated with #32.

**Pull request closed without merging**